### ISSUE TITLE
Fix reward aggregation for multi-step policies

### ISSUE THREAD

**Issue by alexander-soare** [huggingface/lerobot#24]
## Problem

Currently there is an issue with the way `PushtEnv` reports on rewards and success.

`PushtEnv._step` executes multiple steps in the underlying `PushTenv` (note the capitalized T as different from `PushtEnv`: this is the ported DP code). While doing so it **sums** the rewards, and reports on the **last** value of `done`. The consequences of this are:

1. The reported reward is the sum of multiple steps of reward. The "reward" here is meant to be intersection with the target patch, and has no practical interpretation when summed over multiple steps. We could take the mean instead of the sum, but then we wouldn't be reporting the correct final metrics like "avg_max_reward" (see `eval.py`).
2. Perhaps the `done` condition is met in one of the steps, but not on the final step. A simulation which should have finished, doesn't.

*copy-paste of `PushtEnv._step` for reference:*
```python
def _step(self, tensordict: TensorDict):
    td = tensordict
    action = td["action"].numpy()
    # step expects shape=(4,) so we pad if necessary
    # TODO(rcadene): add info["is_success"] and info["success"] ?
    sum_reward = 0

    if action.ndim == 1:
        action = einops.repeat(action, "c -> t c", t=self.frame_skip)
    else:
        if self.frame_skip > 1:
            raise NotImplementedError()

    num_action_steps = action.shape[0]
    for i in range(num_action_steps):
        raw_obs, reward, done, info = self._env.step(action[i])
        sum_reward += reward

        obs = self._format_raw_obs(raw_obs)

        if self.num_prev_obs > 0:
            stacked_obs = {}
            if "image" in obs:
                self._prev_obs_image_queue.append(obs["image"])
                stacked_obs["image"] = torch.stack(list(self._prev_obs_image_queue))
            if "state" in obs:
                self._prev_obs_state_queue.append(obs["state"])
                stacked_obs["state"] = torch.stack(list(self._prev_obs_state_queue))
            obs = stacked_obs

        self.call_rendering_hooks()

    td = TensorDict(
        {
            "observation": TensorDict(obs, batch_size=[]),
            "reward": torch.tensor([sum_reward], dtype=torch.float32),
            # succes and done are true when coverage > self.success_threshold in env
            "done": torch.tensor([done], dtype=torch.bool),
            "success": torch.tensor([done], dtype=torch.bool),
        },
        batch_size=[],
    )
    return td
```

## Quick and dirty fix

We need to properly evaluate the reward metric during training. A quick fix would be to take the summed reward and divide it by the number of steps that were taken to get a mean over that sequence of actions. This would be technically incorrect but close.

## Proposed "proper" fix

There are other ways of going about this but I might suggest absorbing `n_action_steps` logic into the policy's forward method. It is then the policy's responsibility to keep track of the step number it is up to. For example, consider `n_action_steps == 8`. `EnvBase.rollout` calls `policy(tensordict)` in its main loop. Here's what would happen in each step:

```python
policy(tensordict)  # 0. inference is done and an action trajectory `actions` is produced. `action_ix = 0; return self.actions[self.action_ix]`
policy(tensordict)  # 1. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 2. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 3. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 4. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 5. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 6. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 7. `action_ix += 1; return self.actions[self.action_ix]`
policy(tensordict)  # 0. because `action_ix+1 == n_action_steps` we do inference again and reset `action_ix = 0`; `return self.actions[action_ix]`
```

We can have a `policy.reset()` to clear `actions` and `action_ix`.

## Alternative fix

If we don't feel comfortable with giving the policy an action trajectory state, we might consider making a subclass of `torchrl.EnvBase` and overriding `rollout` to gracefully handle multistep policies. This will do the same as `PushtEnv._step` but it won't aggregate the per-step metrics. The other advantage of this is that it handles multi-step policies universally.

**Comment by alexander-soare** [comment]
![Image](https://github.com/Cadene/lerobot/assets/16543381/bc72369e-1148-437c-87b8-adcf72924607)
